### DISCUSSION TITLE
Why are Starlette sessions so basic?

### DISCUSSION THREAD

**Discussion by RamiAwar** [encode/starlette#1801] [Q&A] [answered]
Why can't we include https://github.com/alex-oleshkevich/starsessions in Starlette?

Even the basic functionality in Starlette is missing so much. And while I can contribute a lot of the changes to make sessions as great as they can be, I just can't bring myself to contribute to the Starlette sessions when there exists a much better library out there.

I also think Sessions should be much better documented, here and in FastAPI. JWTs are really overused and overrated, sessions should be a main feature and the goto authentication mechanism for most web apps. But I digress.

Is there a reason why Starsessions isn't part of Starlette?

**Comment by adriangb** [comment, marked as answer]
Disclaimer: I have not used `starsessions` and this response should not be interpreted as a commentary on that package in any way, even if it is referred to by name.

> Is there a reason why Starsessions isn't part of Starlette?

I think there's a lot of things that could be added to Starlette, but the main goal of Starlette is to be a _minimal_ and _composable_ web toolkit, not a complete web framework with all of the batteries included. The fact that it's possible to write an external package like `starsessions` and easily integrate it into Starlette is a testament to the composability Starlette offers. On the minimal side of things, keeping complexity and dependencies to a minimum is really important for the dev team. Adding `starsessions` would bring in dependencies and code complexity. Generally, if it can be implemented as an external package, we'd prefer that to bringing it into Starlette itself. There are things that can't be implemented as an external package (or that implementing as an external package would make the ecosystem really bad) like the Request/Response model/classes.

> Even the basic functionality in Starlette is missing so much. And while I can contribute a lot of the changes to make sessions as great as they can be, I just can't bring myself to contribute to the Starlette sessions when there exists a much better library out there.

This is the whole point: why do we _need_ to add it to Starlette when there is an external package? If the functionality in Starlette satisfies some fraction of users' basic functionality needs, we are happy to point to an external package for more advanced use cases.

I'll also add that we are happy to direct users to an external package when Starlette is lacking a feature or is too basic. We actually have a whole page in our docs dedicated to this: https://www.starlette.io/third-party-packages/

**Reply by adriangb** [reply]
A relevant discussion: https://github.com/encode/starlette/pull/1386#issuecomment-1038941692

**Reply by RamiAwar** [reply]
Yup I see your point. But also, when the Starlette version of things is really below industry standards, it makes me question whether this should be in Starlette at all.

A better solution that I see is not having Sessions be part of Starlette, but having the documentation in Starlette mention starsessions might be better.

**Reply by adriangb** [reply]
I don't disagree with you

**Reply by Kludex** [reply]
That's a good point. 👍

**Reply by RamiAwar** [reply]
I don't mean to offend just in case, I know this is open source, super grateful to everyone who contributed what they have :D

**Reply by Kludex** [reply]
No one is offended.
